pyproject.toml

```
@@ -55,85 +51,135 @@ docs = [
 
 [project.urls]
-"Bug Tracker" = "https://github.com/zarr-developers/zarr-python/issues"
+Issues = "https://github.com/zarr-developers/zarr-python/issues"
 Changelog = "https://zarr.readthedocs.io/en/stable/release.html"
 Discussions = "https://github.com/zarr-developers/zarr-python/discussions"
 Documentation = "https://zarr.readthedocs.io/"
 Homepage = "https://github.com/zarr-developers/zarr-python"
 
-[tool.coverage.report]
-exclude_lines = [
-    "pragma: no cover",
-    "pragma: ${PY_MAJOR_VERSION} no cover",
+[tool.hatch.version]
+path = "src/zarr/__about__.py"
+
+[[tool.hatch.envs.all.matrix]]
+python = ["3.8", "3.9", "3.10", "3.11"]
+
+[tool.hatch.envs.default]
+dependencies = [
+  "coverage[toml]>=6.5",
+  "pytest",
+  "numcodecs[msgpack]"
 ]
 
-[tool.coverage.run]
-omit = [
-    "zarr/meta_v1.py",
-    "bench/compress_normal.py",
+[tool.hatch.envs.default.scripts]
+test = "pytest {args:tests}"
+test-cov = "coverage run -m pytest {args:tests}"
+cov-report = [
+  "- coverage combine",
+  "coverage report",
+]
+cov = [
+  "test-cov",
+  "cov-report",
+]
+
+[tool.hatch.envs.test]
+extra.dependencies = [
+    "lmdb==1.4.1; sys_platform != 'win32'",
+    "ipytree==0.2.2",
+    "ipywidgets==8.1.0",
+    "azure-storage-blob==12.16.0",
+    "redis==5.0.1",
+    "types-redis",
+    "types-setuptools",
+    "pymongo==4.5.0",
+    "coverage",
+    "pytest-cov==4.1.0",
+    "pytest-doctestplus==1.0.0",
+    "pytest-timeout==2.2.0",
+    "h5py==3.10.0",
+    "fsspec==2023.10.0",
+    "s3fs==2023.10.0",
+    "moto[server]>=4.0.8"
 ]
 
-[tool.setuptools]
-packages = ["zarr", "zarr._storage", "zarr.tests"]
-license-files = ["LICENSE.txt"]
+[tool.hatch.envs.test.env-vars]
+ZARR_V3_EXPERIMENTAL_API= "1"
+ZARR_TEST_SHARDING = "1"
 
-[tool.setuptools_scm]
-version_scheme = "guess-next-dev"
-local_scheme = "dirty-tag"
-write_to = "zarr/version.py"
 
-[tool.ruff]
-line-length = 100
-exclude = [
-    ".bzr",
-    ".direnv",
-    ".eggs",
-    ".git",
-    ".mypy_cache",
-    ".nox",
-    ".pants.d",
-    ".ruff_cache",
-    ".venv",
-    "__pypackages__",
-    "_build",
-    "buck-out",
-    "build",
-    "dist",
-    "venv",
-    "docs"
+[tool.hatch.envs.lint]
+detached = true
+dependencies = [
+  "black>=23.1.0",
+  "mypy>=1.0.0",
+  "ruff>=0.0.243",
+]
+
+[tool.hatch.envs.lint.scripts]
+typing = "mypy --install-types --non-interactive {args:src/zarr tests}"
+style = [
+  "ruff {args:.}",
+  "black --check --diff {args:.}",
+]
+fmt = [
+  "black {args:.}",
+  "ruff --fix {args:.}",
+  "style",
+]
+all = [
+  "style",
+  "typing",
 ]
 
 [tool.black]
+target-version = ["py38"]
+line-length = 100
+skip-string-normalization = true
+
+[tool.ruff]
+target-version = "py38"
 line-length = 100
-exclude = '''
-/(
-    \.git
-  | \.mypy_cache
-  | \.venv
-  | _build
-  | buck-out
-  | build
-  | dist
-  | docs
-)/
-'''
+
+[tool.ruff.isort]
+known-first-party = ["zarr"]
+
+[tool.ruff.flake8-tidy-imports]
+ban-relative-imports = "all"
+
+[tool.ruff.per-file-ignores]
```

ruff can now format too!
```suggestion
[tool.ruff.lint.per-file-ignores]
```